### PR TITLE
Don't try to move files to the trashbin outside of user/files

### LINKED ISSUE #16928: Can't migrate encryption keys from cli
Hi,
on my test installation i installed the newest master revision and activated the server side encryption module. (oC 8.1 latest master, php 5.6.10, mariaDB 10.0)
In my logs this warning rises:
Warning	no app in context	Installation is in transit between the old Encryption (ownCloud <= 8.0) and the new encryption. Please enable the "Default encryption module" and run 'occ encryption:migrate'

I followed the warning:
sudo -u USER /path/to/occ encryption:migrate

The cli got following output:
Reorganize system folder structure
PHP Fatal error:  Call to a member function getRelativePath() on a non-object in /path/to/apps/files_trashbin/lib/storage.php on line 123

### PR DESCRIPTION
**Pull request by schiessle**
Don't try to move files to the trashbin outside of user/files

fix https://github.com/owncloud/core/issues/16928

### CODE REVIEW

File: apps/files_trashbin/lib/storage.php [hunk 1]
```diff
@@ -101,6 +106,27 @@ public function rmdir($path) {
...
+	 * @return bool
+	 */
+	protected function shouldMoveToTrash($path){
+		$normalized = Filesystem::normalizePath($this->mountPoint . '/' . $path);
+		$parts = explode('/', $normalized);
+		if (count($parts) < 4) {
+			return false;
+		}
+
+		if ($this->userManager->userExists($parts[1]) && $parts[2] == 'files') {
```
Comment (PVince81) [inline apps/files_trashbin/lib/storage.php]: Will this have a performance impact ? (ex: LDAP)

Can we just rely on the path here ?

Comment (PVince81) [inline apps/files_trashbin/lib/storage.php]: For example when deleting 1000 files or so

**Comment by PVince81** [comment]
@icewind1991 what do you think ?

File: apps/files_trashbin/lib/storage.php (hunk 1 above)
Comment (schiesbn) [inline apps/files_trashbin/lib/storage.php]: the path is relative to the mount point, e.g. for a external storage mounted at "/user/files/smb/" the path to file "test.txt" would be just "test.txt", so there is no way to decide if the file is located in a valid root. You can only decide this if you have the full path by checking if the path starts with '/user/files'. You also need to check explicitly that "user" is a real user, if you would just check for "files" something like "data/files_encryption/files" would be true too.

**Comment by MorrisJobke** [comment]
@icewind1991 :bell:

File: apps/files_trashbin/lib/storage.php (hunk 1 above)
Comment (icewind1991) [inline apps/files_trashbin/lib/storage.php]: Performance impact should be minimal, usermanager does caching

**Comment by icewind1991** [comment]
Looks good :+1:

**Comment by PVince81** [comment]
:+1:

**Comment by PVince81** [comment]
@karlitschek backport request

**Comment by DeepDiver1975** [comment]
backport is here https://github.com/owncloud/core/pull/17679 - THX

**Comment by karlitschek** [comment]
makes sense. please backport

**Pull request merged**